common/remote/sockfile.cpp

```diff
@@ -5637,6 +5869,179 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
+        }
+        catch (IException *)
+        {
+            reply.rewrite(replyPos);
+            reply.append('-');
+            throw;
+        }
+    }
+
+    bool cmdStreamRead(MemoryBuffer & msg, MemoryBuffer & reply, CRemoteClientHandler &client)
+    {
+        // this is an attempt to authenticate when we haven't got authentication turned on
+        if (TF_TRACE_CLIENT_STATS)
+        {
+            StringBuffer s(client.queryPeerName());
+            PROGLOG("Connect from %s",s.str());
+        }
+
+        Owned<IPropertyTree> requestTree = createPTreeFromJSONString(msg.length(), msg.toByteArray());
+
+        /* Example JSON request:
+         * {
+         *  "format" : "xml",
+         *  "cursor" : "1234",
+         *  "node" : {
+         *   "kind" : "diskread",
+         *   "fileName": "examplefilename",
+         *   "keyfilter" : "f1='1    '",
+         *   "choosen" : "5",
+         *   "cursor" : "12345", // cursor handle
+         *   "input" : {
+         *    "f1" : "string5",
+         *    "f2" : "string5"
+         *   },
+         *   "output" : {
+         *    "f2" : "string",
+         *    "f1" : "real"
+         *   }
+         *  }
+         * }
+         *
+         */
+
+        const char *outputFmtStr = requestTree->queryProp("format");
+        int cursorHandle = requestTree->getPropInt("cursor");
+        Owned<IPropertyTree> responseTree; // Used if xml/json
+        OutputFormat outputFormat;
+        if (!outputFmtStr || strieq("xml", outputFmtStr))
+            outputFormat = outFmt_Xml;
+        else if (strieq("json", outputFmtStr))
+            outputFormat = outFmt_Json;
+        else
+            outputFormat = outFmt_Binary;
+        if (outFmt_Binary != outputFormat)
+            responseTree.setown(createPTree("Response"));
+
+        Owned<IRemoteActivity> outputActivity;
+        OpenFileInfo fileInfo;
+        if (!cursorHandle)
+        {
+            // In future this may be passed the request and build a chain of activities and return sink.
+            outputActivity.setown(createOutputActivity(*requestTree));
+
+            MemoryBuffer cursorMb;
+            if (requestTree->getPropBin("cursorBin", cursorMb))
+            {
+                cursorMb.setEndian(BIG_ENDIAN);
+                outputActivity->restoreCursor(cursorMb);
+            }
+
+            StringBuffer requestStr("jsonrequest:");
+            outputActivity->getInfoStr(requestStr);
+            Owned<StringAttrItem> name = new StringAttrItem(requestStr);
+
+            CriticalBlock block(sect);
+            cursorHandle = getNextHandle();
+            client.previdx = client.openFiles.ordinality();
+            client.openFiles.append(OpenFileInfo(cursorHandle, outputActivity, name));
+        }
+        else if (!lookupFileIOHandle(cursorHandle, fileInfo))
+            cursorHandle = NotFound; // challenge response ..
+        else // known handle, continuation
```

Comment:
What happens if the same continuation request is sent to dafilesrv?  E.g., if the cursor is returned to esp, which then uses it on a subsequent call, and when refreshing the window?

Reply:
Currently the test client (testsocket) does not send the serialized cursor continuation, 
*unless* it gets a challenge response from dafilesrv, at which point it makes another request with the full request + the cursor, so a new dafilesrv can continue.

For the example you give, esp could always _not_ send in a cursor handle and always send in the cursor. That would have the effect of allowing refresh and continuation.

However, I think it's better that if any client does send a cursor, dafilesrv uses it.
That way a client with a handle can also send the cursor if it wants to (e.g. on a refresh), or a client can send the handle only (and be challenged if unknown) and continue to stream.